experiment/image-bumper/main.go

```diff
@@ -0,0 +1,211 @@
+	tagRegexp   = regexp.MustCompile(`(v?\d{8}-(?:v\d(?:[.-]\d+)*-g)?[0-9a-f]{6,10}|latest)(-.+)?`)
+	tagCache    = make(map[string]string)
+)
+
+const (
+	imageHostPart  = 1
+	imageImagePart = 2
+	imageTagPart   = 3
+	tagVersionPart = 1
+	tagExtraPart   = 2
+)
+
+type manifest map[string]struct {
+	TimeCreatedMs string   `json:"timeCreatedMs"`
+	Tags          []string `json:"tag"`
+}
+
+func findLatestTag(imageHost, imageName, currentTag string) (string, error) {
+	k := imageHost + "/" + imageName + ":" + currentTag
+	if result, ok := tagCache[k]; ok {
+		return result, nil
+	}
+
+	currentTagParts := tagRegexp.FindStringSubmatch(currentTag)
+	if currentTagParts == nil {
+		return "", fmt.Errorf("couldn't figure out the current tag in %q", currentTag)
+	}
+	if currentTagParts[tagVersionPart] == "latest" {
+		return currentTag, nil
+	}
+
+	resp, err := http.Get("https://" + imageHost + "/v2/" + imageName + "/tags/list")
+	if err != nil {
+		return "", fmt.Errorf("couldn't fetch tag list: %v", err)
+	}
+
+	result := struct {
+		Manifest manifest `json:"manifest"`
+	}{}
+
+	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
+		return "", fmt.Errorf("couldn't parse tag information from registry: %v", err)
+	}
+
+	latestTag, err := pickBestTag(currentTagParts, result.Manifest)
+	if err != nil {
+		return "", err
+	}
+
+	tagCache[k] = latestTag
+
+	return latestTag, nil
+}
+
+func pickBestTag(currentTagParts []string, manifest manifest) (string, error) {
+	// The approach is to find the most recently created image that has the same suffix as the
+	// current tag. However, if we find one called "latest" (with appropriate suffix), we assume
+	// that's the latest regardless of when it was created.
+	var latestTime int64
+	latestTag := ""
+	for _, v := range manifest {
+		bestVariant := ""
+		override := false
+		for _, t := range v.Tags {
+			parts := tagRegexp.FindStringSubmatch(t)
+			if parts == nil {
+				continue
+			}
+			if parts[tagExtraPart] != currentTagParts[tagExtraPart] {
+				continue
+			}
+			if parts[tagVersionPart] == "latest" {
+				override = true
+				continue
+			}
+			if bestVariant == "" || len(t) < len(bestVariant) {
+				bestVariant = t
+			}
+		}
+		if bestVariant == "" {
+			continue
+		}
+		t, err := strconv.ParseInt(v.TimeCreatedMs, 10, 64)
+		if err != nil {
+			return "", fmt.Errorf("couldn't parse timestamp %q: %v", v.TimeCreatedMs, err)
+		}
+		if override || t > latestTime {
+			latestTime = t
+			latestTag = bestVariant
+			if override {
+				break
+			}
+		}
+	}
+
+	if latestTag == "" {
+		return "", fmt.Errorf("failed to find a good tag")
+	}
+
+	return latestTag, nil
+}
+
+func updateFile(path string, imageFilter *regexp.Regexp) error {
+	content, err := ioutil.ReadFile(path)
+	if err != nil {
+		return fmt.Errorf("failed to read %s: %v", path, err)
+	}
+
+	indexes := imageRegexp.FindAllSubmatchIndex(content, -1)
```

Comment:
why not use replace 
https://golang.org/pkg/regexp/#Regexp.ReplaceAll

Reply:
Each match can have a different replacement, and we don't yet know what that is.

`regexp.ReplaceAllFunc` is close, but AFAICT won't give us the submatches, which we want.